pkg/openstack/loadbalancer.go

```diff
@@ -1150,11 +1150,15 @@ func (lbaas *LbaasV2) buildMonitorCreateOpts(svcConf *serviceConfig, port corev1
 	if port.Protocol == corev1.ProtocolUDP {
 		opts.Type = "UDP-CONNECT"
 	}
-	if svcConf.healthCheckNodePort > 0 && lbaas.canUseHTTPMonitor(port) {
-		opts.Type = "HTTP"
-		opts.URLPath = "/healthz"
-		opts.HTTPMethod = "GET"
-		opts.ExpectedCodes = "200"
+	if svcConf.healthCheckNodePort > 0 {
+		if lbaas.canUseHTTPMonitor(port) {
+			opts.Type = "HTTP"
+			opts.URLPath = "/healthz"
+			opts.HTTPMethod = "GET"
+			opts.ExpectedCodes = "200"
+		} else if port.Protocol == corev1.ProtocolTCP {
+			opts.Type = "TCP"
+		}
```

Comment:
I just realized that `opts.Type` here will already be `TCP` as set in line 1145. So… this should just work normally without this change…?

Reply:
if I am reading the code correctly, if kubernetes service protocol is TCP this should work without change?

Reply:
I can confirm this. We currently run an OKD with the cloud-provider forked for OpenShift based on the commit [https://github.com/openshift/cloud-provider-openstack/tree/6b8f88a9c34a541ffaf4b0abf1d8cd1acb55ff9b](6b8f88a) which is equal to this MRs base regarding the changes.

Although the official documentation of OpenStack states that health monitors are not supported (see [limitations](https://docs.openstack.org/networking-ovn/latest/contributor/design/loadbalancer.html#limitations)), the code exists and works. In our OpenStack cluster we do currently lack [a commit fixing an issue with the creation order of the health monitors and the floating ip](https://opendev.org/openstack/ovn-octavia-provider/commit/8eef0beb1d554f5bfb7426723a830033999fbec5). This means we must first create the service of type `LoadBalancer` and after successful creation of said in OpenStack we can add the annotation `loadbalancer.openstack.org/enable-health-monitor: "true"` (or if added via helm, after creation remove and re-add). Otherwise the internal IP of the OpenStack load balancer would correctly redirect to healthy nodes but the floating IP would redirect to every node in the pool.

The only weird thing we encountered was that the OpenStack pool marks nodes as healthy which lack the pod providing the service. In our case we deployed the nginx ingress controller (NIC) which creates a service of type `LoadBalancer` with port 80 and 443. Before knowing about the bug with the creation order both pools contained healthy marked nodes which did not run a NIC pod and if I remember correctly nodes running a NIC pod did not always get marked as healthy. Also contacting the floating IP was mostly rejected which later turned out to be caused by the described bug. Maybe this behaviour was also caused by the wild back and forth switching between options (we were testing how/if this works). After finding out about this bug I configured the NIC productively and freshly deployed it, removed the annotation and added it back after the health monitors have been removed in OpenStack. Now contacting the floating IP with netcat on both ports works repeatedly as expected, although in the pool for the port 80 is a node (`master-2`) which does not run a NIC pod. Opening a connection to `master-2` within the cluster from another node to the allocated `NodePort` for port 80 does result in a rejection, thus the implementation of the cloud provider/kubernetes works as expected and the issue seems to be on the OpenStack side. Also turning off `master-2` does not remove the healthy mark. But removing a pod and rescheduling it on another node results in a correct update (removing the healthy mark on the old node and adding it to the new node).

(Just to be clear, with the paragraph above I did not want to state that something is wrong with the cloud provider or that we need help/explanation, but only what we found out/which problems (probably only on OpenStack side) we have encountered.)

TL;DR: The provider seems to work without this MR and OpenStack successfully creates the health monitors which work as expected although the horizon UI (and CLI) sometimes marks nodes as healthy which are not running the service.

The only thing which should be updated is the documentation in my opinion. The note that the provider does not support OVN health monitors is proved to be wrong and the supported load balancer method (`lb-method`) for OVN should be changed to `SOURCE_IP_PORT`. Currently the documentation states that it must be `SOURCE_IP` which is not a valid option, at least for OVN.
Maybe if the configuration only contains the `lb-provider` set to `ovn` the code should not default to `ROUND_ROBIN` for the `lb-method`, but that would only be a nice to have feature.
Furthermore the documentation states that the only supported `lb-provider`s are `amphora` and `octavia`, but the code also contains `ovn` in the supported load balancer provider list, thus either the documentation is outdated or the code contains `ovn` although it should not.

Reply:
> I can confirm this. We currently run an OKD with the cloud-provider forked for OpenShift based on the commit [https://github.com/openshift/cloud-provider-openstack/tree/6b8f88a9c34a541ffaf4b0abf1d8cd1acb55ff9b](6b8f88a) which is equal to this MRs base regarding the changes.
> 
> Although the official documentation of OpenStack states that health monitors are not supported (see [limitations](https://docs.openstack.org/networking-ovn/latest/contributor/design/loadbalancer.html#limitations)), the code exists and works. In our OpenStack cluster we do currently lack [a commit fixing an issue with the creation order of the health monitors and the floating ip](https://opendev.org/openstack/ovn-octavia-provider/commit/8eef0beb1d554f5bfb7426723a830033999fbec5). This means we must first create the service of type `LoadBalancer` and after successful creation of said in OpenStack we can add the annotation `loadbalancer.openstack.org/enable-health-monitor: "true"` (or if added via helm, after creation remove and re-add). Otherwise the internal IP of the OpenStack load balancer would correctly redirect to healthy nodes but the floating IP would redirect to every node in the pool.

It's this thing, it's fixed: https://bugs.launchpad.net/neutron/+bug/1997418

> The only weird thing we encountered was that the OpenStack pool marks nodes as healthy which lack the pod providing the service. In our case we deployed the nginx ingress controller (NIC) which creates a service of type `LoadBalancer` with port 80 and 443. Before knowing about the bug with the creation order both pools contained healthy marked nodes which did not run a NIC pod and if I remember correctly nodes running a NIC pod did not always get marked as healthy. Also contacting the floating IP was mostly rejected which later turned out to be caused by the described bug. Maybe this behaviour was also caused by the wild back and forth switching between options (we were testing how/if this works). After finding out about this bug I configured the NIC productively and freshly deployed it, removed the annotation and added it back after the health monitors have been removed in OpenStack. Now contacting the floating IP with netcat on both ports works repeatedly as expected, although in the pool for the port 80 is a node (`master-2`) which does not run a NIC pod. Opening a connection to `master-2` within the cluster from another node to the allocated `NodePort` for port 80 does result in a rejection, thus the implementation of the cloud provider/kubernetes works as expected and the issue seems to be on the OpenStack side. Also turning off `master-2` does not remove the healthy mark. But removing a pod and rescheduling it on another node results in a correct update (removing the healthy mark on the old node and adding it to the new node).

Take a look at https://github.com/kubernetes/cloud-provider-openstack/issues/2148 where we discuss this type of issues. TLDR - this is Octavia's implementation detail that members are considered up by default. Not great.

> (Just to be clear, with the paragraph above I did not want to state that something is wrong with the cloud provider or that we need help/explanation, but only what we found out/which problems (probably only on OpenStack side) we have encountered.)
> 
> TL;DR: The provider seems to work without this MR and OpenStack successfully creates the health monitors which work as expected although the horizon UI (and CLI) sometimes marks nodes as healthy which are not running the service.
> 
> The only thing which should be updated is the documentation in my opinion. The note that the provider does not support OVN health monitors is proved to be wrong and the supported load balancer method (`lb-method`) for OVN should be changed to `SOURCE_IP_PORT`. Currently the documentation states that it must be `SOURCE_IP` which is not a valid option, at least for OVN. Maybe if the configuration only contains the `lb-provider` set to `ovn` the code should not default to `ROUND_ROBIN` for the `lb-method`, but that would only be a nice to have feature. Furthermore the documentation states that the only supported `lb-provider`s are `amphora` and `octavia`, but the code also contains `ovn` in the supported load balancer provider list, thus either the documentation is outdated or the code contains `ovn` although it should not.

Would you be so kind to contribute a PR with docs fixes that you mention here?

Reply:
> It's this thing, it's fixed: https://bugs.launchpad.net/neutron/+bug/1997418

Thanks, we know but currently are unable to roll out this fix. I wrote it if someone else has similar problems and for explanation of the following findings.

> Take a look at https://github.com/kubernetes/cloud-provider-openstack/issues/2148 where we discuss this type of issues. TLDR - this is Octavia's implementation detail that members are considered up by default. Not great.

I'm not quite sure if this is related to our problem that the `master-2` stays up no matter what. I know that nodes are initially up and that is not great and learning by reading the issue that nodes will be marked healthy when creating a new one is irritating, but `master-2` is _healthy_ since the creation and that's a few days ago so I do not think those two issues are related or I misunderstood it. The only think I can think off is, that `master-2` does not get checked its health, but I am not sure how to look up when the last health check has been done.

> Would you be so kind to contribute a PR with docs fixes that you mention here?

Of course. Just to be clear I should do the following:
- remove the note that `ovn` does not support the annotation `loadbalancer.openstack.org/enable-health-monitor` within [expose-applications-using-loadbalancer-type-service.md](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/openstack-cloud-controller-manager/expose-applications-using-loadbalancer-type-service.md)
- remove the note that `ovn` has limited scope completely and add a note to the `lb-method` that `ovn` only supports `SOURCE_IP_PORT` within the [using-openstack-cloud-controller-manager.md](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/openstack-cloud-controller-manager/using-openstack-cloud-controller-manager.md)
- add `SOURCE_IP_PORT` to the list of supported `lb-method`s within the [using-openstack-cloud-controller-manager.md](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/openstack-cloud-controller-manager/using-openstack-cloud-controller-manager.md)
  \-
  Please correct me if I am wrong. As far as I know `ROUND_ROBIN`, `LEAST_CONNECTIONS`, and `SOURCE_IP` are valid for `amphora` and `octavia` and `SOURCE_IP_PORT` is only valid (and the only one) for `ovn`, right?
- add `ovn` as supported/tested `lb-provider` within the [using-openstack-cloud-controller-manager.md](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/openstack-cloud-controller-manager/using-openstack-cloud-controller-manager.md)

Did I miss something?

Reply:
> > It's this thing, it's fixed: https://bugs.launchpad.net/neutron/+bug/1997418
> 
> Thanks, we know but currently are unable to roll out this fix. I wrote it if someone else has similar problems and for explanation of the following findings.
> 
> > Take a look at #2148 where we discuss this type of issues. TLDR - this is Octavia's implementation detail that members are considered up by default. Not great.
> 
> I'm not quite sure if this is related to our problem that the `master-2` stays up no matter what. I know that nodes are initially up and that is not great and learning by reading the issue that nodes will be marked healthy when creating a new one is irritating, but `master-2` is _healthy_ since the creation and that's a few days ago so I do not think those two issues are related or I misunderstood it. The only think I can think off is, that `master-2` does not get checked its health, but I am not sure how to look up when the last health check has been done.

Oh, you're right here, sorry! I'm not sure how to debug that. One thing that we've seen is weird behavior with UDP health monitors, maybe you're using UDP there?

> > Would you be so kind to contribute a PR with docs fixes that you mention here?
> 
> Of course. Just to be clear I should do the following:
> 
> * remove the note that `ovn` does not support the annotation `loadbalancer.openstack.org/enable-health-monitor` within [expose-applications-using-loadbalancer-type-service.md](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/openstack-cloud-controller-manager/expose-applications-using-loadbalancer-type-service.md)
> * remove the note that `ovn` has limited scope completely and add a note to the `lb-method` that `ovn` only supports `SOURCE_IP_PORT` within the [using-openstack-cloud-controller-manager.md](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/openstack-cloud-controller-manager/using-openstack-cloud-controller-manager.md)
> * add `SOURCE_IP_PORT` to the list of supported `lb-method`s within the [using-openstack-cloud-controller-manager.md](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/openstack-cloud-controller-manager/using-openstack-cloud-controller-manager.md)
>   -
>   Please correct me if I am wrong. As far as I know `ROUND_ROBIN`, `LEAST_CONNECTIONS`, and `SOURCE_IP` are valid for `amphora` and `octavia` and `SOURCE_IP_PORT` is only valid (and the only one) for `ovn`, right?
> * add `ovn` as supported/tested `lb-provider` within the [using-openstack-cloud-controller-manager.md](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/openstack-cloud-controller-manager/using-openstack-cloud-controller-manager.md)
> 
> Did I miss something?

No, this makes sense, we'll hash out any doubts on the PR itself.

Reply:
> > Oh, you're right here, sorry! I'm not sure how to debug that. One thing that we've seen is weird behavior with UDP health monitors, maybe you're using UDP there?
> 
> No, we deployed a standard Nginx ingress controller thus both pools are TCP.

Sounds like an Octavia issue then, isn't it? Thinking about this more… Maybe your master-2 accepts the connections so that the TCP monitor is fine with it? That would be contrary to what we observed.

> > No, this makes sense, we'll hash out any doubts on the PR itself.
> 
> I have opened the PR #2489 and will look into EasyCLA.

Reply:
As the following test shows the `master-2` does act correct, thus the issue probably is on OpenStack's side.

```sh
[core@prod-k7xrn-master-0 ~]$ nc -zvw 2 prod-k7xrn-master-2 30507 # does NOT run a pod
nc: connect to prod-k7xrn-master-2 (10.0.3.15) port 30507 (tcp) failed: Connection refused
[core@prod-k7xrn-master-0 ~]$ nc -zvw 2 prod-k7xrn-worker-0-9fv4m 30507 # does run a pod
Connection to prod-k7xrn-worker-0-9fv4m (10.0.1.171) 30507 port [tcp/*] succeeded!
```

A screenshot of the members of the port 80 pool (in https://github.com/kubernetes/cloud-provider-openstack/pull/2489#discussion_r1422639496 more images show the state of the load balancer):
![Pool for port 80](https://github.com/kubernetes/cloud-provider-openstack/assets/23337768/9eddbc5a-11e0-426b-99a6-3f39ae531282)